build/lifecycle.go

```diff
@@ -50,16 +57,34 @@ type LifecycleOptions struct {
 }
 
 func (l *Lifecycle) Execute(ctx context.Context, opts LifecycleOptions) error {
-	cacheImage := cache.New(opts.Image, l.docker)
-	l.logger.Verbose("Using cache image %s", style.Symbol(cacheImage.Image()))
+	l.Setup(opts)
+	defer l.Cleanup()
+
+	lifecycle020OrLater := l.hasVersion020OrLater()
+	var buildCache, launchCache Cache
+	if lifecycle020OrLater {
+		buildCache = cache.NewVolumeCache(opts.Image, "build", l.docker)
+		launchCache = cache.NewVolumeCache(opts.Image, "launch", l.docker)
+		l.logger.Verbose("Using build cache volume %s", style.Symbol(buildCache.Name()))
+		l.logger.Verbose("Using launch cache volume %s", style.Symbol(launchCache.Name()))
+	} else {
+		buildCache = cache.NewImageCache(opts.Image, l.docker)
+		l.logger.Verbose("Using build cache image %s", style.Symbol(buildCache.Name()))
+	}
+
 	if opts.ClearCache {
-		if err := cacheImage.Clear(ctx); err != nil {
-			return errors.Wrap(err, "clearing cache")
+		if err := buildCache.Clear(ctx); err != nil {
+			return errors.Wrap(err, "clearing build cache")
+		}
+		l.logger.Verbose("Build cache %s cleared", style.Symbol(buildCache.Name()))
+
+		if lifecycle020OrLater {
+			if err := launchCache.Clear(ctx); err != nil {
```

Comment:
There might be some weird performance implications of this. Even when `--clear-cache` is passed, the `exporter` still reads the metadata from the previous image and will try to reuse layers if the digests match (note: there shouldn't be metadata only layers in the `--clear-cache` case b/c analyze has been skipped). This will make things slow if we clear the `launch-cache`.

One could argue that the launch cache is just a more performant previous image rather than a cache? (maybe a stretch)

Unless we played something like this https://github.com/buildpack/lifecycle/issues/20 which would result in zero layer reuse in the `--clear-cache` case (and which, as an aside, actually fixes an interesting race condition) or did something else to prevent the exporter from trying to reuse layers, I don't think we should clear the launch cache (cc @sclevine).

Reply:
I agree about thinking of launch cache as a for of "previous image" that happens to be performant. We'll make the change.